crond.8
.\" Copyright 1994 Matthew Dillon (dillon@apollo.backplane.com)
.\" May be distributed under the GNU General Public License
.TH CROND 8 "18 June 2026" "3.3-dev"
.SH NAME
crond \- cron daemon (Dillon's Cron)
.SH SYNOPSIS
.B crond [-l#] [-b] [-c directory]
.SH OPTIONS
.B crond
is a background daemon that parses individual crontab files and
executes commands on behalf of the users in question.
.TP 0.5i
.B "-l<loglevel> "
set logging level, default is 5 (LOG_NOTICE).
.TP 0.5i
.B "-b "
run crond in the background.
.TP 0.5i
.B "-c directory "
specify directory containing crontab files.
.TP 0.5i
.B "-s directory "
specify directory containing system-wide crontab files.  By
default, /etc/cron.d is checked.
.SH DESCRIPTION
.B crond 
is responsible for scanning the crontab files and running
their commands at the appropriate time.  The
.B crontab
program communicates with crond through the "cron.update" file
which resides in crontabs directory, usually /var/spool/cron/crontabs.
This is accomplished by appending the filename of the modified or
deleted crontab file to "cron.update" which crond then picks up to
resynchronize or remove its internal representation of the file.
.PP
.B Crond
has a number of built in limitations to reduce the chance of it being
ill-used.  Potentially infinite loops during parsing are dealt with
via a failsafe counter, and user crontabs are generally limited to
256 crontab entries.  crontab lines may not be longer than 1024 
characters, including the newline.  
.PP
Whenever crond must run a job, it first creates a daemon-owned temporary 
file O_EXCL and O_APPEND to store any output, then fork()s and changes 
its user and group permissions to match that of the user the job is being 
run for, then exec's /bin/sh -c to run the job.  The temporary file remains
under the ownership of the daemon to prevent the user from tampering with
it.  Upon job completion, crond verifies the secureness of the mail file
and, if it has been appended to, mails to the file to user.  The
.B sendmail
program is run under the user's uid to prevent mail related security holes.  
Unlike
.B crontab
, the crond program does not leave an open descriptor to the file for the
duration of the job's execution as this might cause crond to run out 
of descriptors.  When
.B crontab
program allows a user to edit his crontab, it copies the crontab to a user
owned file before running the user's prefered editor.  The suid crontab
programs keeps an open descriptor to the file which it later uses to
copy the file back, thereby ensuring the user has not tampered with the
file type.
.PP
.B Crond
always synchronizes to the top of the minute, checking the current time
against the list of possible jobs.  The list is stored such that the
scan goes very quickly, and crond can deal with several thousand entries
without taking any noticable amount of cpu.
.SH AUTHOR
Matthew Dillon (dillon@apollo.backplane.com)




CHANGELOG
git 5.2 1-July-2026

  * cl_Days[] now a zero based array.

  * main.c renamed crond.c

  * protos.h generation now a little more robust

  * crontab now uses 'replacement' method to write edited files.
    This should prevent issues with editors that replace the file
    when saving.

  * simplify and cleanup RunMainLoop:
    * CheckUpdates() no longer called when a resync is necessary.
    * Remove unnecessary short and long sleeps (stime). sleep
      period is now a constant (WAKEUP_INTERVAL) subject to
      early wakeup owing to SIGNALS.

v5.1 12-June-2026

  * Restore missing month check when running jobs. It had been
    removed by a 3rd party commit taken from Jim's master branch.

  * Crontab resyncs will now happen on the hour instead of every 60
    minutes from the time crond was started.

  * 'cron.update' processing now uses a file lock to avoid race
    conditions.

  * defaults for CRONTABS, CRONSTAMPS and CRONMAIL directory locations
    have been changed as follows
       CRONTABS = /var/lib/cron/crontabs
       CRONSTAMPS = /var/lib/cron/cronstamps
       CRONMAIL = /var/spool/cron

    These differ from their historical locations (but are arguably
    more correct and FHS compliant).

    To continue using the historical locations you will need to
    specify them when running 'make', as follows:
       CRONTABS = /var/spool/cron/crontabs
       CRONSTAMPS = /var/spool/cron/cronstamps
       CRONMAIL = /var/spool/cron/mail

  * other code cleanups and minor fixes. see the git log for
    full details.

v5.0 26-May-2026
  * crond will now clear the process environment before running jobs
    when the '-C' option is specified on startup.  This will avoid the
    need to start the daeon via'env -i' in order to get a pristine
    environment.

  * job output will now be stored in a fixed location, configured via
    the CRONMAIL build option, default: /var/spool/cron/mail rather
    than using a dynamically created directory under TMPDIR.  This
    will prevent repeated daemon stops/starts littering /tmp with
    directories.

  * a new sample script, cronmail-limbo, is provided in the extras/
    directory to handle any mailfiles left behind should the daemon be
    stopped while there are running jobs.

  * crontab now includes signal handlers and a locking mechanism to more
    gracefully handle attempts at concurrent updates to the same crontab.

  * Many other fixes and cleanups. See the git log for details.

v4.5  1-May-2011
  * Some cron jobs were running multiple times. Now we make sure not to
    ArmJobs that are already running; and not to resynchronize while jobs are
    running; and to poll the DST setting. (Fixes Arch FS#18681; thanks to Vincent
    Cappe and Paul Gideon Dann for identifying the second issue; and Tilman
    Sauerbeck for identifying the third.)

  * @monthly was wrongly being parsed the same as @yearly (fixes Arch
    FS#19123). Thanks to Peter Johnson, Paul Gideon Dann, and Tilman Sauerbeck.

  * extra/crond.rc: now uses $CROND_ARGS from /etc/conf.d/crond; sample included
    as extra/crond.conf. Suggested by Eric Bélanger.

  * Running `/etc/rc.d/crond start` after startup could leak unwanted
    environment into cronjobs; now we force crond to start in empty env
    (fixes Arch FS#22085). Thanks to Mantas.

  * Also set LOGNAME environment variable in cronjobs. Requested by Michael
    Trunner; fixes Arch FS#18338.

  * extra/crond.logrotate now correctly gets pid from /var/run/crond.pid
    (fixes Arch FS#18039). Thanks to Kay Abendroth, revel, and Chlump Chatkupt.

  * extra/prune-cronstamps now only deletes files, and is formatted as a
    @weekly crontab. Thanks to Alec Moskvin <alecm@gmx.com>.

  * extra/crontab.vim works around an issue where vim's writebackup would
    interfere with crontab's security model (addresses Arch FS#18352).
    Thanks to Armadillo and Simon Williams.

  * Makefile uses $LDFLAGS (fixes Arch FS#23784). Thanks to Kristoffer Tidemann
    and Mike Frysinger.

  * defs.h sets default locations for CRONTABS and CRONSTAMPS beneath /var/spool/cron/,
    as in earlier versions of dcron.

  * Documentation updates.

  * Thanks for testing and feedback: Feifei Jia, Spider.007, Ray Kohler,
    Igor Zakharoff, Edward Hades, and Joe Lightning.

v4.4  17-Jan-2010
  * Finished mailjobs were being left as zombie processes. Fixed.

  * When using crond with logging-to-file, user jobs could only log some
    events if they had write access to the log. Fixed this by having crond
    keep a file descriptor open to the log; also added a SIGHUP handler
    to make crond re-open the logfile. The sample logrotate script now
    sends that signal.

  * More sensible command-line parsing by crontab.

  * Add prune-cronstamps to extra; document extra/*; general improvement
    of README and manpages.

  * Portability improvements, and defs.h now has fuller comments about
    requirements.

  * Makefile improvements: `make` now caches variables for `make install`;
    don't stomp CFLAGS environment variable, and added BINDIR,SBINDIR,MANDIR.

  * Thanks to Juergen Daubert for more testing and suggestions.

v4.3  11-Jan-2010
  * Internal refactoring to make buffer overflow checks
    clearer and portability issues more explicit.

  * Made file argument to -L mandatory; optional args to
    getopt needs GNU extensions.

  * Makefile tweaks. Added CRONTAB_GROUP for `make install`.
    Renamed TIMESTAMPS -> CRONSTAMPS.

  * Thanks to Juergen Daubert for testing and suggestions.

v4.2  11-Jan-2010
  * Makefile tweaks; moved more constants to #defines.

v4.1  10-Jan-2010
  * Fixed bug in parsing some numeric fields in crontabs. (Terminus of range
    wasn't being modded.)

  * Updated Makefile to make it easier to customize timestamps at configure
    time. Also, if LC_TIME is defined when crond runs, we use that instead of
    compiled-in default (for logging to files, to customize syslog output use
    syslog-ng's 'template' command).


crond.c
//  crond.c
//
//  run as root, but NOT setuid root
//
//  Copyright 1994 Matthew Dillon (dillon@apollo.backplane.com)
//  Copyright 2026 Gary Langshaw (gary.langshaw@gmail.com)
//  May be distributed under the GNU General Public License

#include "database.h"
#include "defs.h"
#include "job.h"
#include "subs.h"

#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/ioctl.h>
#include <syslog.h>
#include <time.h>
#include <unistd.h>

#define ONE_HOUR 3600
#define ONE_MINUTE 60

#ifndef RESCAN_INTERVAL
#define RESCAN_INTERVAL ONE_HOUR
#endif

#ifndef WAKEUP_INTERVAL
#define WAKEUP_INTERVAL ONE_MINUTE
#endif

#ifndef CHECKJOBS_INTERVAL
#define CHECKJOBS_INTERVAL 10
#endif

int LogLevel = 5;
int BackgroundOpt = 0;
const char *CDir = CRONTABS;
const char *SCDir = SCRONTABS;
uid_t DaemonUid;
int InSyncFileRoot;

volatile int sig_chld = 0;

void SigHandler(int sig)
{
  switch (sig)
  {
  case SIGCHLD:
    sig_chld = 1;
    break;
  }
}

void RunMainLoop()
{
  time_t t1;
  time_t t2;
  time_t rescan; /* time of last rescan */
  unsigned int stime;

  t1 = time(NULL);
  t1 = t1 - t1 % WAKEUP_INTERVAL;
  rescan = t1 - t1 % RESCAN_INTERVAL;

  for (;;)
  {
    //  synchronize to 1 second after the minute, minimum sleep of 1 second.

    stime = sleep(WAKEUP_INTERVAL + 1 - time(NULL) % WAKEUP_INTERVAL);
    t2 = time(NULL);

    syslog(LOG_DEBUG, "Wakeup(%s): %s",
           ((stime > 0) ? "interrupted" : "scheduled"), ctime(&t2));

    //  Check for disparity: disparities over an hour either way
    //  result in resynchronization.  A reverse-indexed disparity less
    //  then an hour causes us to effectively sleep until we match the
    //  original time (i.e. no re-execution of jobs that have just
    //  been run).  A forward-indexed disparity less then an hour
    //  causes intermediate jobs to be run, but only once in the worst
    //  case.

    if (t2 < t1 - ONE_HOUR || t2 > t1 + ONE_HOUR)
    {
      rescan = t2 - t2 % RESCAN_INTERVAL;
      t1 = t2 - t2 % WAKEUP_INTERVAL;
      syslog(LOG_NOTICE, "time disparity greater than one hour detected.");
    }

    if (stime == 0 && t2 >= t1)
    {
      if (rescan + RESCAN_INTERVAL <= t2)
      {
        //  The directory is rescanned once an hour to deal with any
        //  screwups.
        rescan = t2 - t2 % RESCAN_INTERVAL;
        SynchronizeDir(CDir, NULL, 0);
        SynchronizeDir(SCDir, "root", 0);
      }
      else
      {
        //  The file 'cron.update' is checked to determine new cron
        //  jobs.
        CheckUpdates(CDir, NULL);
        CheckUpdates(SCDir, "root");
      }

      //  when running jobs, the inequality used is greater but
      //  not equal to t1, and less then or equal to t2.
      TestJobs(t1, t2);
      RunJobs();
      t1 = t2;
    }
    CheckJobs();
  }
}

int main(int argc, char **argv)
{
  extern char *optarg;
  int i;
  int opt;
  struct sigaction sa;

  //  parse options

  DaemonUid = getuid();

  while ((opt = getopt(argc, argv, "bc:l:s:")) != -1)
  {
    switch (opt)
    {
    case 'b':
      BackgroundOpt = 1;
      break;
    case 'c':
      CDir = optarg;
      break;
    case 's':
      SCDir = optarg;
      break;
    case 'l':
      LogLevel = strtol(optarg, NULL, 10);
      break;
    default:
      break;
    }
  }

  //  close stdin, stdout, and unused descriptors.
  //  (stderr normally redirected by caller).

  fclose(stdin);
  fclose(stdout);

  i = open("/dev/null", O_RDWR);
  if (i < 0)
  {
    perror("open: /dev/null:");
    exit(1);
  }
  dup2(i, 0);
  dup2(i, 1);

  for (i = 3; i < OPEN_MAX; ++i)
  {
    close(i);
  }

  if (BackgroundOpt == 1)
  {
    int pid;

    pid = fork();

    if (pid < 0)
    {
      perror("fork");
      exit(1);
    }
    if (pid > 0)
      exit(0);

    setsid();
  }

  //  open syslog and set logging level

  openlog("crond", 0, LOG_CRON);
  setlogmask(LOG_UPTO(LogLevel));

  syslog(LOG_NOTICE, "%s " VERSION " GazL, started", argv[0]);

  //  establish a signal handler for SIGCHLD.

  sa.sa_handler = SigHandler;
  sigemptyset(&sa.sa_mask);
  sa.sa_flags = SA_RESTART | SA_NOCLDSTOP;
  if (sigaction(SIGCHLD, &sa, NULL) == -1)
    syslog(LOG_ERR, "failed to establish sig_handler");

  SynchronizeDir(CDir, NULL, 1);
  SynchronizeDir(SCDir, "root", 1);

  RunMainLoop(); /* does not return */

  return 1;
}
